### ISSUE TITLE
Collaborative Deep Learning for Recommender Systems

### ISSUE THREAD

**Issue by Sunnnyyy16** [Sunnnyyy16/Recommendation_Sys_Research#1]
# Abstract

- Collaborative topic regression(CTR) is an appealing recent method addressing sparsity problem, which tightly couples the two components that learn from two different sources of information, but the latent representation learned by CTR is not very effective when the auxiliary information is very sparse.
- This paper proposes a hierarchical Bayesian model called collaborative deep learning(CDL), which jointly performs deep representation learning for the content information and collaborative filtering for the ratings(feedback) matrix.

# Detail
## Introduction

- Due to the abundance of choice in many online services, recommender system(RS) now play an incereasingly significant role
- 3 Types of RS
    
    1) content-based methods
    
    2) collaborative filtering(CF) based methods
    
    3) hybrid methods
    
- CF based’s limitations: The prediction accuracy often drops significantly when the ratings are very sparse.
- hybrid methods
    - loosely coupled methods
        - process the auxiliary information once and then use it to provide features for the CF models.
    - tightly coupled methods
        - Collaborative topic regression(CTR) is a probabilistic graphical model that seamlessly integrates a topic model, latent Dirichlet allocation(LDA), and a model-based CF methods, probabilistic matrix factorization.
- Bayesian formulation of a deep learning model called stacked denoising autoencoder(SDAE)을 사용한 A hierarchical Bayesian model called collaborative deep learning(CDL)을 제안
- The main contribution of this paper
    - CDL can simultaneously extract an effective deep feature representation from content and capture the similarity and implicit relationship between items and users.
    - CF as a more complex target in a probabilistic framework
    - The algorithm for attaining maximum a posteriori(MAP) estimates. A sampling-based algorithm for the Bayesian treatment of CDL, which is a Bayesian generalized version of back-propagation
    - CDL is first hierarchical Bayesian model to bridge the gap between state-of-the-art DL models and RS. CDL can be easily extended to incorporate other auxiliary information to further boost the performance.
    

## Notation and Problem Formulation

$J$ : items (articles or movies)

$I$ : users

$X_c$: content matrix,  SDAE에서는 clean input

$X_0$: SDAE에서 noise corrupted matrix

$X_l$: output of layer l of the SDAE

$R= [R_{ij}]_{IJ}$ : rating matrix, 0 or 1

$W_l, b_l$: weighted matrix, bias vector of layer l

$L$: number of layers

$W^+$: collection of all layers of weight matrices and biases

L/2- layer SDAE corresponds to an L-layer network

### Stacked Denoising Autoencoders

- a feedforward neural network for learning representations (encoding) of the input data by learning to predict the clean input itself in the output.
    
    
![Untitled](https://user-images.githubusercontent.com/98167266/212784453-bce418cb-9888-4bf0-8d52-06633cf3d113.png)

- the hidden layer in the middle(X2) is constrained to be a bottleneck and the input layer and the input layer X0 is a corrupted version of the clean input data
    
![Untitled1](https://user-images.githubusercontent.com/98167266/212784673-5a504736-a03a-457d-b4e9-a5d34ccb9284.png)

    
- SDAE solves the following optimization problem:
    - $min_{W_l,b_l} ||X_c -X_L||^2_F +\lambda \sum_l||W_l||^2_F$
        - clean input과 SDAE 아웃풋 사이의 오차를 최소화하는 weight 행렬과 bias 찾기
        - penalty 부여→ penalty를 더한 거까지 최소화하는 weight 행렬과 bias 찾기
    - $\lambda$: a regularization parameter

### Generalized Bayesian SDAE

- Assume that both the clean input $X_c$ and the corrupted input $X_0$ are observed.
    
   
![Untitled2](https://user-images.githubusercontent.com/98167266/212784700-536628a3-73a7-4485-b26f-473762d1301a.png)

    
 - If $\lambda_s$ goes to infinity, the Gaussian distribution in Equation will become a Dirac delta distribution centered at $\sigma(X_{l-1,j*}W_l+b_l)$, where $\sigma(.)$  is the sigmoid function
 - The first L/2 layers of the network act as an encoder and the last L/2 layers act as a decoder

### Collaborative Deep Learning

- Using the Bayesian SDAE as a component, the generative process of CDL is defined as follow:
    
 
![Untitled3](https://user-images.githubusercontent.com/98167266/212784715-1a2c7b0f-c956-45a4-b5e4-c28769c51f83.png)

    
  - $\lambda_w, \lambda_n,\lambda_u, \lambda_s, \lambda_v$: hyperparameters
  - $C_{ij}$ : confidence parameter similar to that for CTR
       - $C_{ij}=a$ if $R_{ij}=1$
       - $C_{ij}=b$ otherwise
   - Middle layer $X_{L/2}$ serves as a bridge between the ratings and content info.
       - middle layer가 CDL을 효과적인 feature representation을 학습함과 동시에 similarity와 user들과 item들의 implicit relationship 잡아내는 핵심역할을 수행함.
       - the latent offset $\epsilon_j$도 같은 역할을 수행함.

![Untitled4](https://user-images.githubusercontent.com/98167266/212784727-5c996b04-e645-4bba-bd02-ebae6ef71d7a.png)

### Maximum A Posteriori Estimates

본논문에서는 MAP 추정값을 얻기위해 해당 알고리즘을 고안하여 사용함. 

![Untitled5](https://user-images.githubusercontent.com/98167266/212784740-782ad82b-9577-458e-a34e-0df0200fa445.png)

### Prediction

D : the observed test data

![Untitled6](https://user-images.githubusercontent.com/98167266/212784767-74070e32-4926-4eb6-aad9-e97b0d93be7c.png)

![Untitled7](https://user-images.githubusercontent.com/98167266/212784770-3ed7c91e-4a86-47ee-ad10-fec41d4636ca.png)

- For any new item j with no rating in the training data, its offset $\epsilon^*_j$ will be 0

$R^*_{ij}$ 

--> $R_{ij}$ 의 point estimate

### Datasets

- citeulike-a
- citeulike-t
- netflix
    - ratings, movie titles
    - plots of the corresponding movies

### Evaluation Scheme

- 각 유저와 관련된 P 아이템들을 랜덤하게 선택해 training set을 만들고 나머지는 test set으로 한다.
- 모델들 간 spare와 dense 세팅을 알기 위해 P를 1-10까지 설정한다.
- For each value of P, 다르게 랜덤 선택된 training set으로 evaluation을 5번 반복하고 average performance를 기록한다.
- rating 정보는 implicit feedback의 형태로 있기에 performance 측정을 위해 recall 사용

### Baselines and Experimental Settings

- 비교군
    - CMF(Collective Matrix Factorization)
    - SVDFeatures
    - DeepMusic
    - CTR(Collaborative Topic Regression)
    - CDL(Collaborative Deep Learning) : 본 논문의 모델
    
- CDL 실험 세팅
    - a=1, b=0.01, K=50
    - 하이퍼파라미터  $\lambda_u, \lambda_v, \lambda_n, \lambda_w$에 grid search
    - 5 fold validation
    - clean input(X0)에서 corrupted input(Xc)를 얻기위해 masking noise (노이즈 레벨 0.3) 사용
    - SDAE(L>2)에서 1개 이상의 layer에 droupout(0.1) 사용

### Quantitative Comparison

- Sparse setting(P=1)일 때 Performance 결과

![Untitled8](https://user-images.githubusercontent.com/98167266/212784783-bef0c918-48d7-4bbe-bc99-b7d450b2d7cf.png)

- Dense setting(P=10)일 때 Performance 결과

![Untitled9](https://user-images.githubusercontent.com/98167266/212784795-240a5605-8125-4863-bfa0-7eeaf9d7d61e.png)

- mAP for all models in the sparse settings
    - mAP: mean average precision

![Untitled10](https://user-images.githubusercontent.com/98167266/212784818-ad13fdd3-e654-4849-93b8-ab35a428c835.png)

- CDL with different numbers of layers 적용 결과
    
![Untitled11](https://user-images.githubusercontent.com/98167266/212784831-45d15a25-1eb1-42b6-9cf0-e15deab1ca91.png)

    
![Untitled12](https://user-images.githubusercontent.com/98167266/212784837-6b29fc75-c751-4fdb-b8de-1bab19b7b6c5.png)

- citeulike-a 는 layer2, citeulike-t는 layer 3, netflix는 layer 3이 가장 성능 잘 나옴.
    
- Different values of $\lambda_n$ using citeulike-t under the dense setting.
    
![Untitled13](https://user-images.githubusercontent.com/98167266/212784852-c3a829bd-6ba8-4f8e-9df7-61d34047243a.png)

    
 - The prediction performance degrades significantrly as $\lambda_n$ gets very large or very small.
    

### Qualitative Comparison

- citeulike-t에서 2명의 유저예시를 보고 top3 매치 토픽을 사용하여 각 유저의 프로필을 나타내 보았다
- 3layer(L=6)CDL과 CTR이 추천한 top10 articles.
- user1
    - tag recommendation에 관심 많은 컴퓨터 과학자라 가정했는데 CDL은 정확하게 많은 기사들을 추천한 반면, CTR은 social networks에 집중했다.
    - CDL can better understand the key points of the article to make appropriate recommendation accordingly.
- user2
    - From the matched topics returned by both CDL and CTR, user2 might be a researcher on blood flow dynamic theory particularly in the field of medical science.
    - CDL은 정확하게 유저 프로필을 잡아냈고 정확도는 100%인 반면, CTR는 천문학 기사를 대신 추천함.
- With a more effective representation, CDL can capture the key points of articles and the user preferences more accurately.
- It can model the co-occurrence and relations of words better

## Complexity Analysis and Implementation

- computational complexity of updating $u_i$ = $O(K^2J+K^3)$
    - K : the dimensionality of the learned representation
    - J: the number of items
- complexity of $v_j$ = $O(K^2I+K^3+SK_1)$
    - I : the number of users
    - S : the size of the vocabulary
    - K1= is the dimensionality of the output in the first layer
- $O(SK_1)$ : the cost of computing the output of the encoder and it is dominated by the computation of the first layer.
- the total time complexity : $O(JSK_1+K^2J^2+K^2I^2+K^3)$

## Conclusion and Future Work

- jointly performing deep representation learning for the content information and collaborative filtering for the ratings (feedback) matrix
- CDL : First hierarchical Bayesian model to bridge the gap between state-of-the-art deep learning models and RS
- SDAE 말고 CNN 사용할수도 있다.

<br>
This paper summary is just draft. This will be modified :)

CDL paper link: https://arxiv.org/pdf/1409.2944.pdf
